eta/core/features.py

```diff
@@ -25,118 +25,168 @@
+                called when the Featurizer is already started. The default
+                value is True
+            keep_alive: whether to keep the Featurizer alive (i.e. not to call
+                `stop`) after a each `featurize` call
         '''
-        if warn_on_restart and self._started:
-            logger.warning('featurizer start called when already started.')
+        if warn_on_restart and self._is_started:
+            logger.warning("Featurizer.start() called when already started.")
 
-        if self._started:
+        if self._is_started:
             return
 
-        self._started = True
+        self._is_started = True
         self._keep_alive = keep_alive
         self._start()
 
-    def _stop(self):
-        '''Actual stop code that subclasses need to override.  The public stop
-        function will manage state and invoke this one.
+    def _start(self):
+        '''The backend implementation that is called when the public `start`
+        method is called. Subclasses that require startup configuration should
+        override this method.
         '''
         pass
 
     def stop(self):
-        '''Called by featurize after it finishes to handle state management.
-        If the user called start themselves, this should BE called by them as
-        well; the Featurizer is not able to know the process has ended in this
-        case.
+        '''Stop method that handles any necessary cleanup after featurization
+        is complete.
+
+        This method can be explicitly called by users, and, in fact, it must
+        be called by users who called `start` themselves. If `start` was not
+        called manually or `keep_alive` was set to False, then this method will
+        be invoked at the end of each call to `featurize`.
         '''
-        if not self._started:
+        if not self._is_started:
             return
 
         self._stop()
-        self._started = False
+        self._is_started = False
         self._keep_alive = False
 
-    def _featurize(self, data):
-        '''The core feature extraction routine that subclasses need to
-        implement.'''
-        raise NotImplementedError("subclass must implement _featurize().")
+    def _stop(self):
+        '''The backend implementation that is called when the public `stop`
+        method is called. Subclasses that require cleanup after featurization
+        should override this method.
+        '''
+        pass
 
     def featurize(self, data):
-        '''The core feature extraction routine to be called by users of the
-        featurizer.  It appropriately interacts with the featurizer state
-        management.'''
+        '''Featurizes the input data.
+
+        Args:
+            data: the data to featurize
+
+        Returns:
+            the feature vector
+        '''
         self.start(warn_on_restart=False, keep_alive=False)
-        v = self._featurize(data)
+        fv = self._featurize(data)
```

Comment:
2-char variable names are just as bad as 1-char variable names!  And I am fine with either in cases like this.